app/src/main/java/org/dharmaseed/androidapp/PlayTalkActivity.java

```diff
@@ -324,4 +357,151 @@ public void onStopTrackingTouch(SeekBar seekBar) {
+        if (permission != PackageManager.PERMISSION_GRANTED) {
+            // requestPermissions will ask the user for permission asynchronously
+            // and will call onRequestPermissionsResult() with the results
+            ActivityCompat.requestPermissions(
+                    this,
+                    new String[] {Manifest.permission.WRITE_EXTERNAL_STORAGE},
+                    PERMISSIONS_WRITE_EXTERNAL_STORAGE
+            );
+        } else if (permission == PackageManager.PERMISSION_GRANTED) {
+            new DownloadTalkTask().execute(talk);
+        } else {
+            // should never happen
+            Log.w(LOG_TAG, "Permission was " + permission);
+        }
+    }
+
+    @Override
+    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] results) {
+        switch (requestCode) {
+            case PERMISSIONS_WRITE_EXTERNAL_STORAGE:
+                // we asked for and received permission
+                if (results.length > 0 && results[0] == PackageManager.PERMISSION_GRANTED) {
+                    new DownloadTalkTask().execute(talk);
+                }
+                // if we didn't receive permission, don't do anything
+                return;
+        }
+    }
+
+    public void toggleDownloadImage() {
+        ImageButton downloadButton = (ImageButton) findViewById(R.id.download_button);
+        if (talk.isDownloaded()) {
+            Drawable icon = ContextCompat.getDrawable(this, R.drawable.ic_check_circle_green_24dp);
+            downloadButton.setImageDrawable(icon);
+            downloadButton.setOnClickListener(new View.OnClickListener() {
+                @Override
+                public void onClick(View view) {
+                    onDeleteTalkClicked(view);
+                }
+            });
+        } else {
+            Drawable icon = ContextCompat.getDrawable(this, R.drawable.ic_file_download_green_24dp);
+            downloadButton.setImageDrawable(icon);
+            downloadButton.setOnClickListener(new View.OnClickListener() {
+                @Override
+                public void onClick(View view) {
+                    onDownloadButtonClicked(view);
+                }
+            });
+        }
+    }
+
+    public void onDownloadButtonClicked(View view) {
+        Log.d(LOG_TAG, "Downloading talk " + talk.getId());
+        downloadTalk();
+    }
+
+    public void onDeleteTalkClicked(View view) {
+        DialogFragment dialog = new DeleteTalkFragment();
+        dialog.show(getFragmentManager(), "DeleteTalkFragment");
+    }
+
+    public void deleteTalk() {
+        if (!TalkManager.deleteTalk(talk)) {
+            showToast("Unable to delete talk.", Toast.LENGTH_SHORT);
+        } else {
+            dbManager.deleteTalk(talk);
+            showToast("Talk deleted.", Toast.LENGTH_SHORT);
+            toggleDownloadImage();
+            Log.d(LOG_TAG, "Deleted talk " + talk.getId());
+        }
+    }
+
+    public void startLoadingAnim() {
```

Comment:
I like the download animation, that's cool!

Reply:
😄 me too